### ISSUE TITLE
Disallow timers

### ISSUE THREAD

**Issue by ekrapfl** [eslint/eslint#12764]
**Please describe what the rule should do:**
This rule would disallow usage of `setTimeout()` and `setInterval()` in code.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
var interval =  window.setInterval(() => {
      // Do something
} , 2000);
```
```js
window.setTimeout(() => {
    // Do something
} , 500); 
```
```js
var interval = setInterval(() => {
      // Do something
} , 2000);
```
```js
setTimeout(() => {
    // Do something
} , 500); 
```

**Why should this rule be included in ESLint (instead of a plugin)?**
A lot of developers want to avoid using timers in code unless there is a really good reason.  This rule seems to have a pretty common use case, and could really help to make sure people have to think long and hard if there is another way to solve the problem before using a timer in code.

**Are you willing to submit a pull request to implement this rule?**
No

**Comment by g-plane** [comment]
Thanks for this issue.

However, I'm curious about how you conclude that "a lot of developers want to avoid using timers in code".

**Comment by ekrapfl** [comment]
Fair point.  I guess that was too blanket a statement.  Perhaps a better way of stating it is that timers can easily be abused in code, so it could be beneficial to disallow them in general, and force you to think a little harder about it.

**Comment by mdjermanovic** [comment]
I think this is a good use case for rules such as `no-restricted-globals` and `no-restricted-properties`:

```js
rules: {
    "no-restricted-globals": ["error", {
        name: "setInterval",
        message: "Avoid using timers.",
    }, {
        name: "setTimeout",
        message: "Avoid using timers.",
    }],
    "no-restricted-properties": ["error", {
        object: 'window',
        property: "setInterval",
        message: "Avoid using timers.",
    }, {
        object: "window",
        property: "setTimeout",
        message: "Avoid using timers.",
    }],
},
```

**Comment by ilyavolodin** [comment]
ESLint tries to avoid rules that disable language features (such as setInterval, setTimeout). There are always a lot of people who choose to use some of them, and a lot of people who choose not to use some of them. So those types of rules are very controversial, and that's one of the primary reasons why we try to avoid them. As @mdjermanovic noted, we do have a blanket rules that allows you to disable any language feature you want (no-restricted-properties, no-restricted-globals).

**Comment by ekrapfl** [comment]
I was not aware of `no-restricted-properties` and `no-restricted-globals`.
I agree that those seem like they would cover the use cases I was concerned with.

**Comment by kaicataldo** [comment]
Closing since it looks like we're all in agreement that this should be covered by `no-restricted-properties` and `no-restricted-globals`.
